tests/test_retries.py

```diff
@@ -0,0 +1,175 @@
+        allow_module_level=True,
+    )
+
+
+@inlineCallbacks
+def test_retry_once():
+    retries = deque([True, False])
+    items = []
+
+    with MockServer(EchoResource) as server:
+
+        class ItemPage(ItemWebPage):
+            def to_item(self):
+                if retries.popleft():
+                    raise Retry
+                return {"foo": "bar"}
+
+        class TestSpider(Spider):
+            name = "test_spider"
+            start_urls = [server.root_url]
+
+            custom_settings = {
+                "DOWNLOADER_MIDDLEWARES": {
+                    "scrapy_poet.InjectionMiddleware": 543,
+                },
+                "SPIDER_MIDDLEWARES": {
+                    "scrapy_poet.RetrySpiderMiddleware": 275,
+                },
+            }
+
+            def parse(self, response, page: ItemPage):
+                items.append(page.to_item())
+
+        crawler = make_crawler(TestSpider, {})
+        yield crawler.crawl()
+
+    assert items == [{"foo": "bar"}]
+    assert crawler.stats.get_value("downloader/request_count") == 2
+    assert crawler.stats.get_value("retry/count") == 1
+    assert crawler.stats.get_value("retry/reason_count/page_object_retry") == 1
+    assert crawler.stats.get_value("retry/max_reached") is None
+
+
+@inlineCallbacks
+def test_retry_max():
+    # The default value of the RETRY_TIMES Scrapy setting is 2.
+    retries = deque([True, True, False])
+    items = []
+
+    with MockServer(EchoResource) as server:
+
+        class ItemPage(ItemWebPage):
+            def to_item(self):
+                if retries.popleft():
+                    raise Retry
+                return {"foo": "bar"}
+
+        class TestSpider(Spider):
+            name = "test_spider"
+            start_urls = [server.root_url]
+
+            custom_settings = {
+                "DOWNLOADER_MIDDLEWARES": {
+                    "scrapy_poet.InjectionMiddleware": 543,
+                },
+                "SPIDER_MIDDLEWARES": {
+                    "scrapy_poet.RetrySpiderMiddleware": 275,
+                },
+            }
+
+            def parse(self, response, page: ItemPage):
+                items.append(page.to_item())
+
+        crawler = make_crawler(TestSpider, {})
+        yield crawler.crawl()
+
+    assert items == [{"foo": "bar"}]
+    assert crawler.stats.get_value("downloader/request_count") == 3
+    assert crawler.stats.get_value("retry/count") == 2
+    assert crawler.stats.get_value("retry/reason_count/page_object_retry") == 2
+    assert crawler.stats.get_value("retry/max_reached") is None
+
+
+@inlineCallbacks
+def test_retry_exceeded():
+    # The default value of the RETRY_TIMES Scrapy setting is 2.
+    retries = deque([True, True, True])
+    items = []
+
+    with MockServer(EchoResource) as server:
+
+        class ItemPage(ItemWebPage):
+            def to_item(self):
+                if retries.popleft():
+                    raise Retry
+                return {"foo": "bar"}
```

Comment:
What do you think about testing a case where to_item always raises Retry error? We must ensure that scrapy-poet doesn't keep retrying indefinitely. Or is the idea to make sure tests don't hang if this is not working?

Reply:
I believe the test that covers giving up effectively does that, even though the code is similar to the other tests. But I can modify the test to do it for real.